### PR TITLE
191 create profile page

### PR DESCRIPTION
**Pull request by radu-constantin**
Let me know what improvements I can make!

### CODE REVIEW

**Comment by nina-dk** [comment]
Hey, great job incorporating the form on the page! I haven't yet looked into the TS thing you mentioned, but just by looking at the page, here's a couple of improvements I'm seeing:

- On small screens, the edit form should take the whole screen width. It's aligned a little weirdly to the left now. The Cancel button is a little off, too.
- The input headings are a bit too big, to my view.
- I'd rather give the Bio more space to expand, e.g. `11` Col size, than have that much empty space to the right of the page.
- When the user has no first/last name, the Update button is too close to the username.
- We could precede each username with `@` to differentiate it from the actual user name that's right below. I'd rather that was thinner, too, but I'm always in favour of subtler font sizes/weights so feel free to leave that as is. 😁

**Comment by nina-dk** [comment]
You could use something like this for `UpdateUserInput` and `UpdateUserData`:

```ts
export interface UpdateUserInput extends Omit<User, 'dob' | 'password_hash'> {
  dob?: string;
}

export interface UpdateUserData extends Omit<User, 'password_hash'> {}
```

Otherwise, awesome job! I think the form could take up the whole page width even on medium-sized screens now that I'm looking at it. Re. the bio section, if you don't like it expanded that's totally fine, as well. Lastly, we probably want to allow users to update their usernames, too.

File: apps/backend/src/types.ts [hunk 1]
```diff
@@ -66,6 +66,25 @@ export interface CreateUserData {
   lastName?: string;
 }
 
+/**
+ * format of the data sent to `PUT /user/:id` to edit user
+ */
+export type UpdateUserInput = {
```
Comment (nina-dk) [inline apps/backend/src/types.ts:72]: I've added a comment on refactoring those types.

File: apps/frontend/src/components/MainNavigation/MainNav.tsx [hunk 2]
```diff
@@ -67,6 +92,12 @@ const MainNav = () => {
               <Nav.Link href="#pricing" className={styles.navbarCollapseLink}>
                 NOTIFICATIONS
               </Nav.Link>
+              <Nav.Link className={styles.navbarCollapseLink} onClick={() => {
+                  deleteStoredUser()
+                  window.location.reload()
+              }}>
+                LOG OUT
```
Comment (nina-dk) [inline apps/frontend/src/components/MainNavigation/MainNav.tsx:99]: I think "Sign out" is more common for this

File: apps/frontend/src/components/ProfileInfo/ProfileInfo.tsx [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import { UserWithRelatedData } from '@neighborhood/backend/src/types';
```
Comment (nina-dk) [inline apps/frontend/src/components/ProfileInfo/ProfileInfo.tsx:1]: I believe this should be `UserWithoutPasswordHash`.

File: apps/frontend/package.json [hunk 4]
```diff
@@ -21,6 +21,7 @@
     "bootstrap": "^5.3.0",
     "react": "^18.2.0",
     "react-bootstrap": "^2.9.0",
+    "react-date-picker": "^10.5.2",
```
Comment (nina-dk) [inline apps/frontend/package.json:24]: Is this used somewhere?

File: apps/frontend/src/components/DescriptionBox/DescriptionBox.tsx [hunk 5]
```diff
@@ -113,7 +113,7 @@ export default function DescriptionBox({
           ) : null}
         </Col>
       </Row>
-      <Row className="mt-2 mt-md-4">
+      <Row className="mt-1 md-4">
```
Comment (nina-dk) [inline apps/frontend/src/components/DescriptionBox/DescriptionBox.tsx:116]: What did you intend to do here? `md-4` is not a valid bootstrap class.

File: apps/backend/src/utils/helpers.ts [hunk 6]
```diff
@@ -11,4 +11,15 @@ const isString = (value: unknown): value is string => typeof value === 'string';
  */
 const isObject = (input: unknown): input is Object => !!input && (typeof input === 'object');
 
-export { isString, isObject };
\ No newline at end of file
+
+function stringIsValidDate(input: string): boolean;
```
Comment (nina-dk) [inline apps/backend/src/utils/helpers.ts:15]: Ooh, fancy

**Review by nina-dk — CHANGES_REQUESTED** [review]

File: apps/frontend/src/components/MainNavigation/MainNav.tsx (hunk 2 above)
Comment (radu-constantin) [reply apps/frontend/src/components/MainNavigation/MainNav.tsx]: Done!

File: apps/frontend/package.json (hunk 4 above)
Comment (radu-constantin) [reply apps/frontend/package.json:24]: No, initially I experimented with it for the dob field, but I ended up implementing my own date field. Will delete.

File: apps/frontend/src/components/DescriptionBox/DescriptionBox.tsx (hunk 5 above)
Comment (radu-constantin) [reply apps/frontend/src/components/DescriptionBox/DescriptionBox.tsx:116]: I have no idea what I was trying to do. Will remove it.

File: apps/backend/src/utils/helpers.ts (hunk 6 above)
Comment (radu-constantin) [reply apps/backend/src/utils/helpers.ts:15]: Thanks!

File: apps/frontend/src/components/ProfileInfo/ProfileInfo.tsx (hunk 3 above)
Comment (radu-constantin) [reply apps/frontend/src/components/ProfileInfo/ProfileInfo.tsx:1]: You're right. Will modify

File: apps/backend/src/services/userServices.ts [hunk 7]
```diff
@@ -207,9 +207,69 @@ const createUser = async (userData: CreateUserData): Promise<UserWithoutPassword
   return newUserWithoutPasswordHash;
 };
 
+const isUpdateProfileData = (obj: object): obj is UpdateUserInput => {
+  const VALID_PROPS = ['first_name', 'last_name', 'email', 'dob', 'bio'];
+  const props = Object.keys(obj);
+
+  if (props.some((prop) => !VALID_PROPS.includes(prop))) return false;
+  if ('first_name' in obj && typeof obj.first_name !== 'string') return false;
```
[the remark this answers is not in the document]
Comment (radu-constantin) [reply apps/backend/src/services/userServices.ts:215]: dob is optional

**Review by nina-dk — APPROVED** [review]

**Pull request merged**